pkg/secrets-store/nodeserver.go

```diff
@@ -159,104 +173,116 @@ func (ns *nodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublis
-		if _, err := os.Stat(providerBinary); err != nil {
-			log.Errorf("failed to find provider %s, err: %v", providerName, err)
-			return nil, err
-		}
-
-		parametersStr, err := json.Marshal(parameters)
-		if err != nil {
-			log.Errorf("failed to marshal parameters, err: %v", err)
-			return nil, err
-		}
-		secretStr, err := json.Marshal(secrets)
-		if err != nil {
-			log.Errorf("failed to marshal secrets, err: %v", err)
-			return nil, err
-		}
-		permissionStr, err := json.Marshal(permission)
-		if err != nil {
-			log.Errorf("failed to marshal file permission, err: %v", err)
-			return nil, err
-		}
-
-		// mount before providers can write content to it
-		err = ns.mounter.Mount("tmpfs", targetPath, "tmpfs", []string{})
+	if isMockProvider(providerName) {
+		// mock provider is used only for running sanity tests against the driver
+		err := ns.mounter.Mount("tmpfs", targetPath, "tmpfs", []string{})
 		if err != nil {
 			log.Errorf("mount err: %v", err)
 			return nil, err
 		}
+		log.Infof("skipping calling provider as its mock")
+		return &csi.NodePublishVolumeResponse{}, nil
+	}
 
-		log.Debugf("Calling provider: %s", providerName)
-
-		// check if minimum compatible provider version with current driver version is set
-		// if minimum version is not provided, skip check
-		if _, exists := ns.minProviderVersions[providerName]; !exists {
-			log.Warningf("minimum compatible %s provider version not set", providerName)
-		} else {
-			// check if provider is compatible with driver
-			providerCompatible, err := version.IsProviderCompatible(providerBinary, ns.minProviderVersions[providerName])
-			if err != nil {
-				return nil, err
-			}
-			if !providerCompatible {
-				return nil, fmt.Errorf("Minimum supported %s provider version with current driver is %s", providerName, ns.minProviderVersions[providerName])
-			}
-		}
-
-		args := []string{
-			"--attributes", string(parametersStr),
-			"--secrets", string(secretStr),
-			"--targetPath", string(targetPath),
-			"--permission", string(permissionStr),
-		}
+	// ensure it's read-only
+	if !req.GetReadonly() {
+		return nil, status.Error(codes.InvalidArgument, "Readonly is not true in request")
+	}
+	// get provider volume path
+	providerVolumePath := ns.providerVolumePath
+	if providerVolumePath == "" {
+		return nil, fmt.Errorf("Providers volume path not found. Set PROVIDERS_VOLUME_PATH")
+	}
 
-		log.Infof("provider command invoked: %s %s %v", providerBinary,
-			"--attributes [REDACTED] --secrets [REDACTED]", args[4:])
+	providerBinary := ns.getProviderPath(runtime.GOOS, providerName)
+	if _, err := os.Stat(providerBinary); err != nil {
+		errorReason = ProviderBinaryNotFound
+		log.Errorf("failed to find provider %s, err: %v", providerName, err)
+		return nil, err
+	}
 
-		cmd := exec.Command(
-			providerBinary,
-			args...,
-		)
+	parametersStr, err := json.Marshal(parameters)
+	if err != nil {
+		log.Errorf("failed to marshal parameters, err: %v", err)
```

Comment:
should we also return this as another errorReason?

Reply:
I think these would categorize as `InternalError` rather than individual errors for each marshal error. WDYT?